### DISCUSSION TITLE
Synchronization issues with RealTimeScheduler

### DISCUSSION THREAD

**Discussion by zhangteng0526** [omnetpp/omnetpp#989] [Q&A] [answered]
Thank you for allowing us to discuss some INET issues here. I have another question as follows: I want to implement network simulation (OMNeT++ as a network emulator receives the position information passed in the gazebo of the physical simulator), so I use ExtLowerEthernetInterface, and the use of this port requires the RealTimescheduler scheduling class. Although the role of RealTimescheduler is to simulate the time as close as possible to the time in the real world. Makes the simulation appear to be running in real time, but this still has some lag. How to erase or reduce this difference. get your help, thank you! ! !

ps: gazebo is time-driven

**Comment by rhornig** [comment, marked as answer]
Simulation will always have some kind of extra overhead and you will get lag especially if you are running the simulation under Qtenv (the whole GUI introduces a lot of delay). On the other hand, you seem to want to create a co-simulation where two simulator run in tandem. None of the simulators are inherently 'real-time' so using the ExtLowerEtherInterface is a bit hard constraint. To get the best possible result you would need to write your own Scheduler class that can synchronize between the gazebo and omnetpp simulation kernels. That way you would be able to speed up gazebo and omnetpp would follow or vice versa. I'm not sure if gazebo has an API that allows souch a tight control for its simulation loop, but if it has, I would go along that route and leave out the realtime requirement.

**Reply by zhangteng0526** [reply]
OK, thanks. Can I understand this way? Under the premise of the RealTimeScheduler scheduling class, there is no fundamental way to solve this delay. There seems to be only one way to reduce this latency (reduce emulation events). In addition, in express mode, does setting realtimescheduler-scaling have no effect or meaningless,

**Reply by rhornig** [reply]
- generally, running a realtime scheduler under Qtenv is menaingless as Qtenv can stop or slow down the simulation run which obviously will not stop the real time progressing :) so the whole simulation gets out of sync and violates the realtime scheduler contract. As "express mode" is just a mode of Qtenv, this pretty much implies what you are saying (i.e. the realtimescheduler-scaling is menaingless under Qtenv on whatever mode).
- the delay you are seeing can be reduced only by: a) using faster CPU (single core performance), b) less events

And the correct solution for this problem is to create a scheduler specifically for gazebo and OMNeT++ avoiding the requirement of real-time synchronization.

**Reply by zhangteng0526** [reply]
Thanks, I finally figured it out! This problem has been solved, I will try to write my own scheduling class. In addition, I raised a new problem, which has troubled me for a long time, I hope you can help me find the problem and suggest a solution. Thanks again
